File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    start = 1
    end = times[-1] * n
    
    while start <= end:
        mid = (start + end) // 2
         
        total = 0
        for time in times:
            total += mid // time
            
        if total >= n:
            end = mid - 1
        else:
            start = mid + 1
    
    return start

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        
        long start = 1;
        long end = (long)times[times.length-1] * n;
        
        while (start < end) {
            long mid = ((long)start + end) / 2;
            
            long total = 0;
            for (int time: times) {
                total += mid / time;
            }
            
            if (total >= n) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }
        return end;
    }
}
